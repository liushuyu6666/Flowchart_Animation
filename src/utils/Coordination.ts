export interface Coordination {
    x: number;
    y: number;
}

// export class CoordinationImpl {
//     coordination: Coordination;

//     constructor(x: number, y: number) {
//         this.coordination = {
//             x: x,
//             y: y
//         };
//     }

//     public calculateRelative(row: number, col: number): Coordination {
//         return {
//             x: this.coordination.x - row,
//             y: this.coordination.y - col,
//         };
//     }
// }