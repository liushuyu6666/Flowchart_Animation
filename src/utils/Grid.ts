export interface GridArea {
    rowStart: number;
    rowEnd: number;
    columnStart: number;
    columnEnd: number;
}

export interface GridTemplate {
    gridTemplateRows: number;
    gridTemplateColumns: number;
}