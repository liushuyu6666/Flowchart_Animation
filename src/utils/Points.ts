import { Coordination } from './Coordination';
import { GridArea, GridTemplate } from './Grid';

export class Points {
    points: Coordination[];
    gridArea: GridArea | undefined;
    gridTemplate: GridTemplate | undefined;

    constructor() {
        this.points = [];
        this.gridArea = undefined;
        this.gridTemplate = undefined;
    }

    public addPoint(point: Coordination) {
        this.points.push(point);

        if (!this.gridArea || !this.gridTemplate) {
            // Update girdArea.
            this.gridArea = {
                rowStart: point.x,
                rowEnd: point.x,
                columnStart: point.y,
                columnEnd: point.y,
            };

            // Update gridTemplate.
            this.gridTemplate = {
                gridTemplateRows: 0,
                gridTemplateColumns: 0,
            };
        } else {
            // Update girdArea.
            if (point.x < this.gridArea.rowStart) {
                this.gridArea.rowStart = point.x;
            }
            if (point.x > this.gridArea.rowEnd) {
                this.gridArea.rowEnd = point.x;
            }
            if (point.y < this.gridArea.columnStart) {
                this.gridArea.columnStart = point.y;
            }
            if (point.y > this.gridArea.columnEnd) {
                this.gridArea.columnEnd = point.y;
            }

            // Update gridTemplate.
            this.gridTemplate.gridTemplateRows = Math.abs(
                this.gridArea!.rowEnd - this.gridArea!.rowStart,
            );
            this.gridTemplate.gridTemplateColumns = Math.abs(
                this.gridArea!.columnEnd - this.gridArea!.columnStart,
            );
        }
    }
}
