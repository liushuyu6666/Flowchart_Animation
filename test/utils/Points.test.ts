import { Coordination } from '../../src/utils/Coordination';
import { Points } from '../../src/utils/Points';

describe('Points', () => {
    let points: Points;

    beforeEach(() => {
        points = new Points();
    });

    describe('addPoint', () => {
        test('should update gridArea and gridTemplate when adding the first point', () => {
            const point: Coordination = { x: 2, y: 3 };

            points.addPoint(point);

            expect(points.points).toMatchSnapshot();
            expect(points.gridArea).toMatchSnapshot();
            expect(points.gridTemplate).toMatchSnapshot();
        });

        test('should update gridArea and gridTemplate when adding subsequent points', () => {
            points.addPoint({ x: 2, y: 3 });
            points.addPoint({ x: 4, y: 1 });
            points.addPoint({ x: 3, y: 5 });

            expect(points.points).toMatchSnapshot();
            expect(points.gridArea).toMatchSnapshot();
            expect(points.gridTemplate).toMatchSnapshot();
        });
    });
});
