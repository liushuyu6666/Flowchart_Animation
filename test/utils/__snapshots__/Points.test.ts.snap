// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Points addPoint should update gridArea and gridTemplate when adding subsequent points 1`] = `
[
  {
    "x": 2,
    "y": 3,
  },
  {
    "x": 4,
    "y": 1,
  },
  {
    "x": 3,
    "y": 5,
  },
]
`;

exports[`Points addPoint should update gridArea and gridTemplate when adding subsequent points 2`] = `
{
  "columnEnd": 5,
  "columnStart": 1,
  "rowEnd": 4,
  "rowStart": 2,
}
`;

exports[`Points addPoint should update gridArea and gridTemplate when adding subsequent points 3`] = `
{
  "gridTemplateColumns": 4,
  "gridTemplateRows": 2,
}
`;

exports[`Points addPoint should update gridArea and gridTemplate when adding the first point 1`] = `
[
  {
    "x": 2,
    "y": 3,
  },
]
`;

exports[`Points addPoint should update gridArea and gridTemplate when adding the first point 2`] = `
{
  "columnEnd": 3,
  "columnStart": 3,
  "rowEnd": 2,
  "rowStart": 2,
}
`;

exports[`Points addPoint should update gridArea and gridTemplate when adding the first point 3`] = `
{
  "gridTemplateColumns": 0,
  "gridTemplateRows": 0,
}
`;
